Product Collection: Fix "Products to show" inspector control not showing as expected on Cart template

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #52171

This PR addresses an issue where the "Products to show" accordion item in the inspector was empty when adding an "Upsells" Collection to the Cart template. The fix ensures that the correct UI and options are displayed in the inspector, allowing users to choose between showing upsells from products in the cart or from a specific product.


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

**Case 1: Products to show accordion is visible with correct UI**

1. Go to `Page: Cart` template.
2. Add a Product Collection block and choose "Upsells" collection.
3. Verify that "PRODUCTS TO SHOW" accordion is visible with the correct UI as shown in screenshot below
![image](https://github.com/user-attachments/assets/410b8c2c-f24e-445b-bba3-3231ef2e8c01)

4. Verify that by default, "From products in the cart" option is selected.

**Case 2: Testing on Cart page**

1. Go to `Page: Cart` template.
2. Add a Product Collection block and choose "Upsells" collection.
3. Publish the page and go to frontend
4. Add a product with upsells to cart
5. Go to cart page
6. Verify that:
	- Collection is rendered with "You may also like" heading
	- Collection shows correct upsells products from products which are in cart
	- For more info, see PR #51374
7. Go back to Editor and change "From products in the cart" to "From a specific product" 
8. Select a specific product and save the page
9. Go to frontend and verify that on Cart page, upsells products are always shown from a specific product you selected in the inspector control. These products should not change when you add/remove other products to cart.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Product Collection: Fix "Products to show" inspector control not showing as expected on Cart template
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Fix: Upsells collection not showing correct products on frontend

This commit addresses an issue where the Upsells collection wasn't displaying the correct products on the frontend. The problem was caused by a variable name mismatch in the condition checking for empty product references.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

imanish003 committed:
Rename product_references variable for consistency

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into fix/upsells-collection-issue

imanish003 committed:
Add support for linked products in cart context

This commit extends the LinkedProductControl component to support
linked products in the cart context. It introduces the following changes:
- Add a check for cart location alongside product location
- Introduce a dynamic label for the radio control option, which changes
  based on whether the context is a product or cart
- Refactor the logic to determine when to show the radio control and
  which option should be selected by default

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

imanish003 committed:
Refactor LinkedProductControl for improved clarity and flexibility

- Extract getFromCurrentProductRadioLabel function for better reusability
- Improve conditional logic for showing linked product control
- Enhance readability by using more descriptive variable names
- Simplify showLinkedProductControl calculation

imanish003 committed:
Merge branch 'fix/52171' of https://github.com/woocommerce/woocommerce into fix/52171

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -136,35 +147,40 @@ const LinkedProductControl = ( {
 	location: WooCommerceBlockLocation;
 	usesReference: string[] | undefined;
 } ) => {
-	const REFERENCE_TYPE_PRODUCT = 'product';
```
I think it would be a good idea to keep the `const` instead of using the string ('product'). Using the const is less error prone. We could also create another constant for 'cart'.

imanish003 replied:
Makes sense to me. As we were using 'product' in only two places, therefore I removed the constant and replaced it with string. But Let's use constants for both 🤝

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -136,35 +147,40 @@ const LinkedProductControl = ( {
-	const isProductReferenceAvailable =
-		location.type === REFERENCE_TYPE_PRODUCT;
+	const isProductLocation = location.type === 'product';
+	const isUsesReferenceIncludesProduct =
```
Totally nitpicky; can we rename this const to something like: `hasProductReference`?
I just want something smaller and more descriptive.

imanish003 replied:
I don't have a strong opinion on this, so made the changes in d6de3a2

octaedro commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx:
```diff
@@ -136,35 +147,40 @@ const LinkedProductControl = ( {
+	const isUsesReferenceIncludesProduct =
+		!! usesReference?.includes( 'product' );
+	const isCartLocation = location.type === 'cart';
+	const isUsesReferenceIncludesCart = !! usesReference?.includes( 'cart' );
```
Same here; what do you think about renaming this const to `hasCartReference`?

imanish003 replied:
I don't have a strong opinion on this, so I made the changes in d6de3a2

octaedro reviewed (commented):
Good job @imanish003! This is testing well on my end and the code looks great! I just left a couple of small (and optional) comments.
I also see that we should remove [this useMemo](https://github.com/woocommerce/woocommerce/blob/2b9044a0a6b0ce70bd83f911b8c8cfaea83f82b0/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx#L8), in order to make the [linter pass](https://github.com/woocommerce/woocommerce/actions/runs/11439424020/job/31823103997?pr=52186#step:4:37).

imanish003 committed:
Refactor: Add constants for reference types in LinkedProductControl

imanish003 committed:
Rename variables in LinkedProductControl for clarity

- 'isUsesReferenceIncludesProduct' to 'hasProductReference'
- 'isUsesReferenceIncludesCart' to 'hasCartReference'

imanish003 committed:
Remove unused import

imanish003 commented:
> I also see that we should remove [this useMemo](https://github.com/woocommerce/woocommerce/blob/2b9044a0a6b0ce70bd83f911b8c8cfaea83f82b0/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/linked-product-control.tsx#L8), in order to make the [linter pass](https://github.com/woocommerce/woocommerce/actions/runs/11439424020/job/31823103997?pr=52186#step:4:37).

@octaedro Good catch! Made the changes in [f77f3ed](https://github.com/woocommerce/woocommerce/pull/52186/commits/f77f3ed3a65d5e5b19e1ed6559de525d9de27d40).

imanish003 committed:
Merge branch 'trunk' into fix/52171

imanish003 committed:
Revert changes in ProductCollection.php

octaedro reviewed (approved):
Thank you @imanish003 for addressing the changes. You did a fantastic job here 🚀
